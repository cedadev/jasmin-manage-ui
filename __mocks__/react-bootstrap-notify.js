export const useNotifications = jest.fn(() => jest.fn());