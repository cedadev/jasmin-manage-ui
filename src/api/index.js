import { useEffect, useState } from 'react';

import {
    apiFetch,
    createContextForEndpoint,
    createContextForResource,
    useResource,
    useInstance,
    useFetchPoint
} from '../rest-resource';


// Create a context for the current user and export the hook
const CurrentUser = createContextForEndpoint("/api/me/");
const CurrentUserProvider = CurrentUser.Provider;
export const useCurrentUser = CurrentUser.hook;


// All other resources require authentication before they can be fetched
// This higher-order component wraps a provider so that it is only mounted once
// the user is successfully initialised
const withAuthentication = Provider => ({ children }) => {
    const currentUser = useCurrentUser();
    // Only wrap the children in the provider if the user is authenticated
    return currentUser.initialised ? <Provider>{children}</Provider> : children;
};


// Create a context each for the top-level, read-only resources that don't change
// often and we want to share across the component tree
const Categories = createContextForResource("/api/categories/");
const Consortia = createContextForResource("/api/consortia/");
const Resources = createContextForResource("/api/resources/");

// Wrap the providers with the higher-order component for authentication
const CategoriesProvider = withAuthentication(Categories.Provider);
const ConsortiaProvider = withAuthentication(Consortia.Provider);
const ResourcesProvider = withAuthentication(Resources.Provider);

// Export the hooks for each resource
export const useCategories = Categories.hook;
export const useConsortia = Consortia.hook;
export const useResources = Resources.hook;

// Export a single provider that provides all the top-level resources
export const Provider = ({ children }) => (
    <CurrentUserProvider>
        <CategoriesProvider>
            <ConsortiaProvider>
                <ResourcesProvider>
                    {children}
                </ResourcesProvider>
            </ConsortiaProvider>
        </CategoriesProvider>
    </CurrentUserProvider>
);


// Export a hook for loading a consortium by id
export const useConsortium = (id, options) => useInstance(`/api/consortia/${id}/`, options);

// Export hooks for accessing projects
// We don't treat the project list as a top-level shared resource because:
//   1. The list, or objects within it, changes relatively often
//   2. Projects can be loaded by id that are not in the list, e.g. when the user
//      is a consortium manager, which reduces the value in caching the list
export const useProjects = options => useResource("/api/projects/", options);
export const useProject = (id, options) => useInstance(`/api/projects/${id}/`, options);

// Export hook for loading requirement by id
export const useRequirement = (id, options) => useInstance(`/api/requirements/${id}/`, options);

// Export hook for loading service by id
export const useService = (id, options) => useInstance(`/api/services/${id}/`, options);

// Export a function for joining a project with an invite code
export const joinProject = async code => apiFetch('/api/join/', { method: 'POST', data: { code } });

// Export hook for loading category by id
export const useCategory = (id, options) => useInstance(`/api/categories/${id}/`, options);

// Export hook for loading resource by id
export const useResourceType = (id, options) => useInstance(`/api/resources/${id}/`, options);


// Export a hook for using a project event stream
export const useProjectEvents = (project, options) => {
    const { fetchPoint = true, autoFetch = true } = options || {};

    const [initialised, setInitialised] = useState(false);
    const [dirty, setDirty] = useState(autoFetch);
    const [fetching, setFetching] = useState(false);
    const [fetchError, setFetchError] = useState(null);
    const [data, setData] = useState([]);

    // Get the events URL from the project links
    const eventsUrl = project.data._links.events;

    // Build the stream object with fetchable methods
    const stream = {
        initialised,
        dirty,
        fetching,
        fetchError,
        data,
        // This function loads more events
        fetch: () => {
            // Build the URL to use
            const url = new URL(eventsUrl);
            // If there is a most recent event, use its time for the since parameter
            const event = data.find(_ => true);
            if( event ) url.searchParams.append("since", event.created_at);
            // Mark the fetch as in progress
            setFetching(true);
            // In order to prevent state updates after a component has been unmounted,
            // we need to be able to cancel fetches
            // To do this, we can use an abort controller
            const controller = new AbortController();
            // Start the fetch using the signal from the controller
            apiFetch(url, { signal: controller.signal })
                // If the promise resolves, update the state with the new data
                .then(newData => {
                    // Prepend the data to the existing data
                    setData([...newData, ...data]);
                    setInitialised(true);
                    setFetching(false);
                    setDirty(false);
                    setFetchError(null);
                })
                // If the promise rejects, update the state with the error
                .catch(error => {
                    // If the error is an abort error, it is because the component that
                    // owns the data has been unmounted so don't update the state
                    if( error.name === "AbortError" ) return;
                    setFetching(false);
                    setDirty(false);
                    setFetchError(error);
                });
            // Return the abort function to allow the fetch to be cancelled from outside
            return () => controller.abort();
        },
        // This function marks the stream as dirty, causing more data to be fetched
        markDirty: () => setDirty(true),
        // This function resets the stream, clearing all the data and triggering a re-fetch
        reset: () => {
            setInitialised(false);
            setDirty(true);
            setFetching(false);
            setFetchError(null);
            setData([]);
        }
    };

    // Hooks cannot be conditional, so we always have to call the fetch point hook
    // However we can effectively disable it by making sure it never sees a dirty stream
    useFetchPoint({ ...stream, dirty: fetchPoint && stream.dirty });

    // We need to be a fetch point for the event stream
    return stream;
};
