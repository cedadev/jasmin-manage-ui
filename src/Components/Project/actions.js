import { useNestedResource, useAggregateResource } from '../../rest-resource';

import { useConsortia, useCurrentUser } from '../../api';


/**
 * Hook to access the current user's roles for the given project.
 */
export const useProjectUserRoles = project => {
    const currentUser = useCurrentUser();
    // Extract the consortium for the project
    // If the consortia are not loaded, return nothing
    const consortia = useConsortia();
    const consortium = consortia.data[project.data.consortium];
    // This is not a fetch point for collaborators - we just want to use the
    // information if it has been loaded elsewhere
    const collaborators = useNestedResource(project, "collaborators", { fetchPoint: false });

    // Get the user's role, either from the collaborator record (if one exists) or from the project
    const userCollaborator = Object.values(collaborators.data)
        .filter(c => c.data.user.id === currentUser.data.id)
        .shift();
    const userRole = userCollaborator ?
        userCollaborator.data.role :
        project.data.current_user_role;

    return {
        // If the user has a role, they are a collaborator
        isProjectCollaborator: !!userRole,
        // If the user role is OWNER, then they are an owner
        isProjectOwner: userRole && userRole === "OWNER",
        // For this, we need to check the consortium
        isConsortiumManager: (
            // Wait for the consortia to load before deciding
            consortium &&
            consortium.data.manager.id === currentUser.data.id
        )
    };
};


/**
 * Hook to access the current user's permissions for the given project.
 */
export const useProjectPermissions = project => {
    const projectRoles = useProjectUserRoles(project);
    return {
        // The project can be edited by collaborators when in editable mode
        canEditRequirements: (
            project.data.status === "EDITABLE" &&
            projectRoles.isProjectCollaborator
        ),
        // The project can be submitted for review by a project owner
        // when in editable mode
        canSubmitForReview: (
            project.data.status === "EDITABLE" &&
            projectRoles.isProjectOwner
        ),
        // Requirements can be approved by consortium managers, but only when
        // the project is under review
        canApproveRequirements: (
            project.data.status === "UNDER_REVIEW" &&
            projectRoles.isConsortiumManager
        ),
        // Changes can be requested by the consortium manager for a project
        // that is under review
        canRequestChanges: (
            project.data.status === "UNDER_REVIEW" &&
            projectRoles.isConsortiumManager
        ),
        // A project that is under review can be submitted for provisioning by
        // the consortium manager
        canSubmitForProvisioning: (
            project.data.status === "UNDER_REVIEW" &&
            projectRoles.isConsortiumManager
        )
    };
};


/**
 * Hook to access the available actions for the current project.
 */
export const useProjectActions = project => {
    // Get all the project requirements by aggregating across the services
    // However this is not a fetch point for services - we only use them once
    // they have been loaded elsewhere in the component tree
    const services = useNestedResource(project, "services", { fetchPoint: false });
    const requirements = useAggregateResource(services, "requirements");

    // Work out if the project has requirements with various statuses
    const hasRequirementsWithStatus = status => {
        const count = Object.values(requirements.data)
            .filter(requirement => requirement.data.status === status)
            .length;
        return count > 0;
    };
    const hasRequestedRequirements = hasRequirementsWithStatus("REQUESTED");
    const hasRejectedRequirements = hasRequirementsWithStatus("REJECTED");
    const hasApprovedRequirements = hasRequirementsWithStatus("APPROVED");

    // All these actions are not allowed while the project is updating
    return {
        // Requirements cannot be edited while the project is updating
        allowEditRequirements: !project.updating,
        // The project can only be submitted for review when there are no
        // rejected requirements and at least one requirement in either the
        // requested or approved state
        // That is to cover for the case where a requirement was rejected
        // alongside others that were approved, and then subsequently removed
        allowSubmitForReview: (
            !project.updating &&
            !hasRejectedRequirements &&
            (hasRequestedRequirements || hasApprovedRequirements)
        ),
        // Requirements cannot be approved while the project is updating
        allowApproveRequirements: !project.updating,
        // Changes can be requested for a project only when:
        //   1. All requested requirements have been resolved
        //   2. There is at least one rejected requirement
        allowRequestChanges: (
            !project.updating &&
            !hasRequestedRequirements &&
            hasRejectedRequirements
        ),
        // A project can be submitted for provisioning only once:
        //   1. All the requested requirements have been resolved
        //   2. There are no rejected requirements
        //   3. There is at least one approved requirement
        allowSubmitForProvisioning: (
            !project.updating &&
            !hasRequestedRequirements &&
            !hasRejectedRequirements &&
            hasApprovedRequirements
        )
    };
};
