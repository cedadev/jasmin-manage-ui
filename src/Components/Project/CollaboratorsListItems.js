import React, { useState } from 'react';

import Badge from 'react-bootstrap/Badge';
import Col from 'react-bootstrap/Col';
import FormControl from 'react-bootstrap/FormControl';
import ListGroup from 'react-bootstrap/ListGroup';
import Modal from 'react-bootstrap/Modal';
import Row from 'react-bootstrap/Row';

import classNames from 'classnames';

import moment from 'moment';

import ReactSelect from 'react-select';

import { useNotifications } from 'react-bootstrap-notify';

import { InstanceDeleteButton } from '../../rest-resource';

import { useCurrentUser } from '../../api';

import { notificationFromError, sortByKey } from '../utils';

import { useProjectPermissions } from './actions';

import '../../css/stacked-modals.css';


const collaboratorRoles = ["CONTRIBUTOR", "OWNER"];


/**
 * Renders a select that allows the collaborator role to be changed.
 */
const CollaboratorRoleSelect = ({ collaborator, className, disabled, ...props }) => {
    const notify = useNotifications();

    const [selectedRole, setSelectedRole] = useState(collaborator.data.role);
    const options = collaboratorRoles.map(role => ({ value: role, label: role }));
    const selectedOption = options.find(opt => opt.value === selectedRole);

    const handleChange = async option => {
        const role = option.value;
        // Store the updated value for the time being
        setSelectedRole(role);
        // Try to update the contributor
        try {
            await collaborator.update({ role });
        }
        catch(error) {
            // On error, produce a notification and reset the value
            notify(notificationFromError(error));
            setSelectedRole(collaborator.data.role);
        }
    }

    return (
        <ReactSelect
            {...props}
            className={classNames("rich-select", className)}
            options={options}
            value={selectedOption}
            onChange={handleChange}
            isDisabled={collaborator.updating || disabled}
            // This is required to make sure the menu goes over the
            // invite button, which has z-index: 2
            styles={{ menu: styles => ({ ...styles, zIndex: 3 }) }}
        />
    );
};


const CollaboratorDeleteButton = ({ collaborator, disabled }) => {
    const notify = useNotifications();

    const collaboratorName = collaborator.data.user.last_name ?
        `${collaborator.data.user.first_name} ${collaborator.data.user.last_name}` :
        collaborator.data.username;

     // Handle a delete error by producing a notification
     const handleError = error => notify(notificationFromError(error));

    return (
        <InstanceDeleteButton
            instance={collaborator}
            onError={handleError}
            disabled={disabled}
        >
            <Modal.Header>
                <Modal.Title>Delete collaborator</Modal.Title>
            </Modal.Header>
            <Modal.Body>
                <p>
                    Are you sure you want to remove{" "}
                    <strong>{collaboratorName}</strong>{" "}
                    as a collaborator?
                </p>
                <p className="mb-0">
                    <strong>
                        Once removed, a collaborator can only rejoin the project by invitation.
                    </strong>
                </p>
            </Modal.Body>
        </InstanceDeleteButton>
    )
};


const CollaboratorUserDetails = ({ collaborator, displayName }) => {
    const currentUser = useCurrentUser();
    const createdAt = moment(collaborator.data.created_at).fromNow();
    return (<>
        <p className="mb-1">
            {displayName(collaborator)}
            {collaborator.data.user.id === currentUser.data.id && (
                <Badge variant="warning" style={{ fontSize: '90%' }} className="ml-2">
                    You
                </Badge>
            )}
        </p>
        <p className="mb-0">
            <small className="text-muted">Added {createdAt}</small>
        </p>
    </>);
};


const CollaboratorListItemNoEdit = ({ collaborator, displayName }) => (
    <ListGroup.Item>
        <div className="d-flex align-items-center">
            <div className="flex-grow-1">
                <CollaboratorUserDetails
                    collaborator={collaborator}
                    displayName={displayName}
                />
            </div>
            <div>
                <Badge
                    variant={collaborator.data.role === "OWNER" ? "danger" : "success"}
                    style={{ fontSize: '100%' }}
                    className="px-3 py-2"
                >
                    {collaborator.data.role}
                </Badge>
            </div>
        </div>
    </ListGroup.Item>
);


const CollaboratorListItemEdit = ({ collaborator, displayName, disabled }) => (
    <ListGroup.Item className="py-2">
        <Row className="align-items-center">
            <Col xs={12} lg={7}>
                <CollaboratorUserDetails
                    collaborator={collaborator}
                    displayName={displayName}
                />
            </Col>
            <Col xs={12} lg={5}>
                <div className="d-flex">
                    <FormControl
                        as={CollaboratorRoleSelect}
                        collaborator={collaborator}
                        className="mr-2"
                        disabled={disabled}
                    />
                    <CollaboratorDeleteButton
                        collaborator={collaborator}
                        disabled={disabled}
                    />
                </div>
            </Col>
        </Row>
    </ListGroup.Item>
);


export const CollaboratorsListItems = ({ project, collaborators }) => {
    // See if the current user is allowed to edit the collaborators
    const { canEditCollaborators } = useProjectPermissions(project);
    const CollaboratorListItemComponent = canEditCollaborators ?
        CollaboratorListItemEdit :
        CollaboratorListItemNoEdit;

    // Check if the project has multiple owners
    // If there is only only one owner, we want to disable the deleting or changing of the owner
    const numOwners = Object.values(collaborators.data)
        .filter(c => c.data.role === "OWNER")
        .length;
    const soleOwner = numOwners === 1;

    // Sort the collaborators by display name, using the username as a fallback
    const displayName = c => c.data.user.last_name ?
        `${c.data.user.first_name} ${c.data.user.last_name} (${c.data.user.username})` :
        c.data.user.username;
    const sortedCollaborators = sortByKey(Object.values(collaborators.data), displayName);

    return sortedCollaborators.map(collaborator => (
        <CollaboratorListItemComponent
            key={collaborator.data.id}
            collaborator={collaborator}
            displayName={displayName}
            disabled={soleOwner && collaborator.data.role === "OWNER"}
        />
    ));
};
