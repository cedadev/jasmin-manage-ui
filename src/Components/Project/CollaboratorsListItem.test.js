import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { CollaboratorsListItems } from "./CollaboratorsListItems";
import { useProjectPermissions } from "./actions";
import { useCurrentUser } from "../../api";
import { useNotifications } from "react-bootstrap-notify";
import { act } from "react-dom/test-utils";
import { useHistory } from "react-router-dom";
import renderer from "react-test-renderer";

// Mock the dependencies
jest.mock("moment", () => {
  const mockMoment = () => ({
    fromNow: () => "a few days ago",
  });
  return mockMoment;
});

jest.mock("./actions", () => ({
  useProjectPermissions: jest.fn(),
}));

jest.mock("../../api", () => ({
  useCurrentUser: jest.fn(),
}));

jest.mock("../../rest-resource", () => ({
  InstanceDeleteButton: ({ children, disabled, onSuccess, onError }) => (
    <div>
      <button
        onClick={() => {
          if (onSuccess) onSuccess();
          if (onError) onError(new Error("Delete failed"));
        }}
        disabled={disabled}
        data-testid="delete-button"
      >
        Delete collaborator
      </button>
      {children}
    </div>
  ),
}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: jest.fn(),
}));

jest.mock("react-select", () => {
  const ReactSelect = ({ options, value, onChange, isDisabled }) => {
    const handleChange = (e) => {
      const selectedOption = options.find(
        (opt) => opt.value === e.target.value
      );
      onChange(selectedOption);
    };

    return (
      <select
        data-testid="role-select"
        value={value?.value || ""}
        onChange={handleChange}
        disabled={isDisabled}
      >
        {options.map((opt) => (
          <option key={opt.value} value={opt.value}>
            {opt.label}
          </option>
        ))}
      </select>
    );
  };
  return ReactSelect;
});

describe("CollaboratorsListItems Components", () => {
  const mockProject = { id: 1, name: "Test Project", data: { consortium: 1 } };
  const mockCollaborators = {
    data: {
      1: {
        data: {
          id: 1,
          user: {
            id: 1,
            username: "user1",
            first_name: "User",
            last_name: "One",
          },
          role: "OWNER",
          created_at: "2025-01-06T00:00:00Z",
        },
      },
      2: {
        data: {
          id: 2,
          user: {
            id: 2,
            username: "user2",
            first_name: "User",
            last_name: "Two",
          },
          role: "CONTRIBUTOR",
          created_at: "2025-01-05T00:00:00Z",
        },
      },
    },
  };

  const mockNotify = jest.fn();
  useNotifications.mockReturnValue(mockNotify);

  beforeEach(() => {
    jest.clearAllMocks();
  });

  const renderCollaboratorsListItems = (permissions, currentUser) => {
    useProjectPermissions.mockReturnValue(permissions);
    useCurrentUser.mockReturnValue(currentUser);

    render(
      <Router>
        <CollaboratorsListItems
          project={mockProject}
          collaborators={mockCollaborators}
        />
      </Router>
    );
  };

  describe("CollaboratorsListItems", () => {
    it("Should match the snapshot", () => {
      const component = renderer.create(
        renderCollaboratorsListItems(
          { canEditCollaborators: true },
          { data: { id: 1 } },
          mockCollaborators
        )
      );

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("Should render a list of collaborators with edit permissions", () => {
      renderCollaboratorsListItems(
        { canEditCollaborators: true },
        { data: { id: 1 } }
      );

      // Check that the collaborators are rendered
      expect(screen.getByText("User One (user1)")).toBeInTheDocument();
      expect(screen.getByText("User Two (user2)")).toBeInTheDocument();

      // Check that the role select and delete button are rendered
      expect(screen.getAllByRole("combobox")).toHaveLength(2);
      expect(
        screen.getAllByRole("button", { name: /Delete collaborator/i })
      ).toHaveLength(2);
    });

    it("Should render a list of collaborators without edit permissions", () => {
      renderCollaboratorsListItems(
        { canEditCollaborators: false },
        { data: { id: 1 } }
      );

      // Check that the collaborators are rendered
      expect(screen.getByText("User One (user1)")).toBeInTheDocument();
      expect(screen.getByText("User Two (user2)")).toBeInTheDocument();

      // Check that the role select and delete button are not rendered
      expect(screen.queryByRole("combobox")).not.toBeInTheDocument();
      expect(
        screen.queryByRole("button", { name: /delete/i })
      ).not.toBeInTheDocument();
    });

    it("Should disable delete button for sole owner", () => {
      renderCollaboratorsListItems(
        { canEditCollaborators: true },
        { data: { id: 1 } }
      );

      // Check that the delete button for the sole owner is disabled
      const deleteButtons = screen.getAllByRole("button", { name: /delete/i });

      expect(deleteButtons[0]).toBeDisabled();
      expect(deleteButtons[1]).not.toBeDisabled();
    });

    it("Should render collaborator details correctly", () => {
      renderCollaboratorsListItems(
        { canEditCollaborators: true },
        { data: { id: 1 } }
      );

      // Check that the collaborator details are rendered correctly
      expect(screen.getByText("User One (user1)")).toBeInTheDocument();
      expect(screen.getByText("User Two (user2)")).toBeInTheDocument();
      expect(screen.getAllByText("Added a few days ago")).toHaveLength(2);
    });

    it("Should render collaborator role select correctly", () => {
      renderCollaboratorsListItems(
        { canEditCollaborators: true },
        { data: { id: 1 } }
      );

      // Check that the role select is rendered correctly
      const roleSelects = screen.getAllByRole("combobox");
      expect(roleSelects[0]).toHaveDisplayValue("OWNER");
      expect(roleSelects[1]).toHaveDisplayValue("CONTRIBUTOR");
    });

    it("Should render delete button correctly", () => {
      renderCollaboratorsListItems(
        { canEditCollaborators: true },
        { data: { id: 1 } }
      );

      // Check that the delete button is rendered correctly
      const deleteButtons = screen.getAllByRole("button", { name: /delete/i });
      expect(deleteButtons).toHaveLength(2);
    });

    it("Should handle role change successfully", async () => {
      const mockNotify = jest.fn();
      useNotifications.mockReturnValue(mockNotify);

      // Setup mock data with update function
      const mockCollaborator = {
        ...mockCollaborators.data[1],
        data: {
          ...mockCollaborators.data[1].data,
          role: "CONTRIBUTOR",
        },
        update: jest.fn().mockResolvedValue({}),
      };

      renderCollaboratorsListItems(
        { canEditCollaborators: true },
        { data: { id: 1 } }
      );

      // Find role select by its text content
      const ownerRole = screen.getByDisplayValue("OWNER");
      const contributorRole = screen.getByDisplayValue("CONTRIBUTOR");

      expect(ownerRole).toBeInTheDocument();
      expect(contributorRole).toBeInTheDocument();

      // Change the role
      await act(async () => {
        fireEvent.change(contributorRole, { target: { value: "OWNER" } });
      });

      // Verify only one OWNER is displayed since the sole owner can't be changed
      expect(screen.getAllByDisplayValue("OWNER")).toHaveLength(1);
      expect(screen.getByDisplayValue("CONTRIBUTOR")).toBeInTheDocument();
    });

    it("Should handle error on role update", async () => {
      const mockNotify = jest.fn();
      useNotifications.mockReturnValue(mockNotify);

      // Create mock collaborators with an error-throwing update function
      const mockCollaboratorsWithError = {
        data: {
          1: {
            ...mockCollaborators.data[1],
            update: jest.fn().mockRejectedValue(new Error("Update failed")),
          },
          2: mockCollaborators.data[2],
        },
      };

      render(
        <Router>
          <CollaboratorsListItems
            project={mockProject}
            collaborators={mockCollaboratorsWithError}
          />
        </Router>
      );

      // Find and change the role select
      const roleSelect = screen.getAllByRole("combobox")[0];
      await act(async () => {
        fireEvent.change(roleSelect, { target: { value: "OWNER" } });
      });

      // Verify the error notification was shown
      await waitFor(() => {
        expect(mockNotify).toHaveBeenCalled();
      });
    });

    it("Should redirect to projects page when user deletes themselves", async () => {
      const history = { push: jest.fn() };
      useHistory.mockReturnValue(history);
      useCurrentUser.mockReturnValue({ data: { id: 1 } });

      const mockCollaborators = {
        data: {
          1: {
            data: {
              id: 1,
              user: {
                id: 1,
                username: "user1",
                first_name: "User",
                last_name: "One",
              },
              role: "OWNER",
              created_at: "2025-01-06T00:00:00Z",
            },
          },
          2: {
            data: {
              id: 2,
              user: {
                id: 2,
                username: "user2",
                first_name: "User",
                last_name: "Two",
              },
              role: "OWNER",
              created_at: "2025-01-05T00:00:00Z",
            },
          },
          3: {
            data: {
              id: 3,
              user: {
                id: 3,
                username: "user3",
                first_name: "User",
                last_name: "Three",
              },
              role: "CONTRIBUTOR",
              created_at: "2025-01-04T00:00:00Z",
            },
          },
        },
      };

      render(
        <Router>
          <CollaboratorsListItems
            project={mockProject}
            collaborators={mockCollaborators}
          />
        </Router>
      );

      const deleteButton = screen.getAllByRole("button", {
        name: /delete/i,
      })[0];

      await act(async () => {
        fireEvent.click(deleteButton);
      });

      await waitFor(() => {
        expect(history.push).toHaveBeenCalledWith("/projects");
      });
    });

    const renderCollaboratorsListItems = (
      permissions,
      currentUser,
      collaborators = mockCollaborators
    ) => {
      useProjectPermissions.mockReturnValue(permissions);
      useCurrentUser.mockReturnValue(currentUser);

      render(
        <Router>
          <CollaboratorsListItems
            project={mockProject}
            collaborators={collaborators}
          />
        </Router>
      );
    };

    it("Should display username when last name is not available", () => {
      const mockCollaboratorsNoLastName = {
        data: {
          1: {
            data: {
              id: 1,
              user: {
                id: 1,
                username: "testuser",
                first_name: "Test",
              },
              role: "OWNER",
              created_at: "2023-01-01T00:00:00Z",
            },
          },
        },
      };

      renderCollaboratorsListItems(
        { canEditCollaborators: true },
        { data: { id: 1 } },
        mockCollaboratorsNoLastName
      );

      expect(screen.getByText("testuser")).toBeInTheDocument();
    });

    it("Should handle error on collaborator deletion", async () => {
      const mockNotify = jest.fn();
      useNotifications.mockReturnValue(mockNotify);

      const mockCollaboratorsWithDeleteError = {
        data: {
          1: {
            ...mockCollaborators.data[1],
            delete: jest.fn().mockRejectedValue(new Error("Delete failed")),
            role: "CONTRIBUTOR",
            deleting: false,
            updating: false,
            data: {
              id: 1,
              user: {
                id: 1,
                username: "user1",
                first_name: "User",
              },
              role: "CONTRIBUTOR",
              created_at: "2023-01-01T00:00:00Z",
            },
          },
        },
      };

      renderCollaboratorsListItems(
        { canEditCollaborators: true },
        { data: { id: 1 } },
        mockCollaboratorsWithDeleteError
      );

      // Click initial delete button to open modal
      const deleteButton = screen.getByTestId("delete-button");
      fireEvent.click(deleteButton);

      // Find and click the confirm delete button in the modal footer
      const confirmDeleteButton = screen.getByRole("button", {
        name: /delete/i,
      });
      fireEvent.click(confirmDeleteButton);

      // Verify notification was shown
      await waitFor(() => {
        expect(mockNotify).toHaveBeenCalledWith(
          expect.objectContaining({
            message: expect.stringContaining("Delete failed"),
          })
        );
      });
    });

    it("Should disable role select when collaborator is updating", () => {
      const mockCollaboratorsUpdating = {
        data: {
          1: {
            data: {
              id: 1,
              user: {
                id: 1,
                username: "user1",
                first_name: "User",
                last_name: "One",
              },
              role: "OWNER",
              created_at: "2023-01-01T00:00:00Z",
            },
            updating: true,
          },
        },
      };

      renderCollaboratorsListItems(
        { canEditCollaborators: true },
        { data: { id: 1 } },
        mockCollaboratorsUpdating
      );

      const roleSelect = screen.getByRole("combobox");
      expect(roleSelect).toBeDisabled();
    });

    it("Should handle deletion of another user without redirect", async () => {
      const history = { push: jest.fn() };
      useHistory.mockReturnValue(history);

      // Mock current user with different ID than collaborator
      useCurrentUser.mockReturnValue({ data: { id: 999 } });

      const mockCollaboratorsOtherUser = {
        data: {
          1: {
            ...mockCollaborators.data[1],
            delete: jest.fn().mockResolvedValue({}),
            data: {
              id: 1,
              user: {
                id: 2, // Different from current user id
                username: "other",
                first_name: "Other",
              },
              role: "CONTRIBUTOR",
              created_at: "2023-01-01T00:00:00Z",
            },
          },
        },
      };

      renderCollaboratorsListItems(
        { canEditCollaborators: true },
        { data: { id: 1 } },
        mockCollaboratorsOtherUser
      );

      // Click delete and confirm
      const deleteButton = screen.getByTestId("delete-button");
      fireEvent.click(deleteButton);

      const confirmButton = screen.getByRole("button", { name: /delete/i });
      fireEvent.click(confirmButton);

      // Verify no redirect happened
      await waitFor(() => {
        expect(history.push).not.toHaveBeenCalled();
      });
    });
  });
});
