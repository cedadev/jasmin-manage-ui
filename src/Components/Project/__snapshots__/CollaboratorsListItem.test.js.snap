// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollaboratorsListItems Components CollaboratorsListItems Should match the snapshot 1`] = `null`;
